pymbolic/primitives.py

```diff
@@ -535,13 +559,24 @@ def __repr__(self):
 
     # {{{ hash/equality interface
 
+    # This custom warning deduplication mechanism became necessary because the
+    # sheer amount of warnings ended up leading to out-of-memory situations
+    # with pytest which bufered all the warnings.
```

Comment:
Nice, this is an interesting mechanism!

As a side note, at least in my tests with #147, the slowdown did not seem to be caused by out-of-memory situations; the memory consumed stayed well below the memory size of the machine.

So I'd suggest changing the text to:
```suggestion
    # This custom warning deduplication mechanism became necessary because the
    # sheer number of warnings ended up leading to slow performance
    # with pytest which buffered all the warnings.
```

Reply:
If that was the case, why were the runs canceled? More explicitly, [this CI run](https://github.com/inducer/pymbolic/actions/runs/11001495578) explicitly states:

> The hosted runner: GitHub Actions 4 lost communication with the server. Anything in your workflow that terminates the runner process, starves it for CPU/Memory, or blocks its network access can cause this error.

Reply:
With the pyfmmlib update, I was able to run the majority (~70%) of tests locally, and it seems like it is a memory usage issue after all. The final tally is ~38 GByte RSS HWM, with 24.5 million warnings.

The main culprit appears to be `test_symbolic.py::test_derivative_binder_expr` (29 GByte RSS, 19.5M warnings):
```
test/test_symbolic.py: 6513704 warnings
  /shared/home/mdiener/Work/emirge/pymbolic/pymbolic/primitives.py:574: DeprecationWarning: Expression.__hash__ is used by <class 'pytential.symbolic.primitives.NodeCoordinateComponent'>. This is deprecated. Use hash functions supplied by augment_expression_dataclass instead. This will stop working in 2025.
    elif hash(self) != hash(other):

test/test_symbolic.py: 3240040 warnings
  <dataclass augmentation code for <class 'pymbolic.primitives.CommonSubexpression'>>:20: DeprecationWarning: Expression.__eq__ is used by <class 'pytential.symbolic.primitives.NumReferenceDerivative'>. This is deprecated. Use equality comparison supplied by augment_expression_dataclass instead. This will stop working in 2025.

test/test_symbolic.py: 6513632 warnings
  /shared/home/mdiener/Work/emirge/pymbolic/pymbolic/primitives.py:574: DeprecationWarning: Expression.__hash__ is used by <class 'pytential.symbolic.primitives.NumReferenceDerivative'>. This is deprecated. Use hash functions supplied by augment_expression_dataclass instead. This will stop working in 2025.
    elif hash(self) != hash(other):

test/test_symbolic.py: 3256816 warnings
  /shared/home/mdiener/Work/emirge/pymbolic/pymbolic/primitives.py:612: DeprecationWarning: Expression.__eq__ is used by <class 'pytential.symbolic.primitives.NodeCoordinateComponent'>. This is deprecated. Use equality comparison supplied by augment_expression_dataclass instead. This will stop working in 2025.
    and self.__getinitargs__() == other.__getinitargs__())
```

 Another interesting point is that 50% of the RSS happens after the test has finished, possibly from some post-processing of warnings.